### PR TITLE
DOC: Clarify the type alias deprecation message

### PR DESCRIPTION
**Pull request by seberg**
This tries to clarify the type alias deprecation message slightly
to give more guidance on how to review the replacement (if desired).

---

<details> <summary> The new messages are as follows: </summary>

```
In [1]: np.bool                                                                                                                                             
<ipython-input-1-190120470286>:1: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.bool
Out[1]: bool

In [2]: np.object                                                                                                                                           
<ipython-input-2-4921af820668>:1: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.object
Out[2]: object

In [3]: np.str                                                                                                                                              
<ipython-input-3-15a72efb409b>:1: DeprecationWarning: `np.str` is a deprecated alias for the builtin `str`. To silence this warning, use `str` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.str_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.str
Out[3]: str

In [4]: np.float                                                                                                                                            
<ipython-input-4-b3b46df38e48>:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.float
Out[4]: float

In [5]: np.complex                                                                                                                                          
<ipython-input-5-087dad13a955>:1: DeprecationWarning: `np.complex` is a deprecated alias for the builtin `complex`. To silence this warning, use `complex` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.complex128` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.complex
Out[5]: complex

In [6]: np.int                                                                                                                                              
<ipython-input-6-5dcbb6b27796>:1: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.int
Out[6]: int

In [7]: np.long                                                                                                                                             
<ipython-input-7-2567cec2f44b>:1: DeprecationWarning: `np.long` is a deprecated alias for `np.compat.long`. To silence this warning, use `np.compat.long` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `int` for which `np.compat.long` is itself an alias. Doing this will not modify any behaviour and is safe. When replacing `np.long`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.long
Out[7]: int

In [8]: np.unicode                                                                                                                                          
<ipython-input-8-a16bff60575c>:1: DeprecationWarning: `np.unciode` is a deprecated alias for `np.compat.unciode`. To silence this warning, use `np.compat.unciode` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `str` for which `np.compat.unciode` is itself an alias. Doing this will not modify any behaviour and is safe. If you specifically wanted the numpy scalar type, use `np.str_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.unicode
```

### CODE REVIEW

**Comment by seberg** [comment]
I tried to give some more details about integers in the release notes, and mention the precision version `float64` rather than `float_` in the message.  So hopefully it is a bit more clear, it may be slightly opinionated.

**Comment by charris** [comment]
What was the reason to change the 1.20.0 release note?

File: doc/source/release/1.20.0-notes.rst [hunk 1]
```diff
@@ -75,32 +75,59 @@ Using the aliases of builtin types like ``np.int`` is deprecated
...
+The third column lists alternative NumPy names which may occasionally be
+preferential. See also `user/basics.types`_ for additional details.
+
+=================  ============  ==================================================================
+Deprecated name    Identical to  NumPy scalar type names
+=================  ============  ==================================================================
+``numpy.bool``     ``bool``      `numpy.bool_`
+``numpy.int``      ``int``       `numpy.int_` (default), ``numpy.int64``, or ``numpy.int32``
+``numpy.float``    ``float``     `numpy.float64`, `numpy.float_`, `numpy.double` (equivalent)
+``numpy.complex``  ``complex``   `numpy.complex128`, ``numpy.complex_`, `numpy.cdouble` (equivalent)
```
Comment (charris) [inline doc/source/release/1.20.0-notes.rst:95]: Needs backtick  `numpy.complex_`. Or maybe fewer :)

File: doc/source/release/1.20.0-notes.rst [hunk 2]
```diff
@@ -75,32 +75,59 @@ Using the aliases of builtin types like ``np.int`` is deprecated
...
-``numpy.object``    ``object``                         `numpy.object_`
-``numpy.str``       ``str``                            `numpy.str_`
-``numpy.long``      ``int`` (``long`` on Python 2)     `numpy.int_` (C ``long``), `numpy.longlong` (largest integer type)
-``numpy.unicode``   ``str`` (``unicode`` on Python 2)  `numpy.unicode_`
-==================  =================================  ==================================================================
-
-Note that for technical reasons these deprecation warnings will only be emitted
-on Python 3.7 and above.
+The third column lists alternative NumPy names which may occasionally be
+preferential. See also `user/basics.types`_ for additional details.
```
Comment (charris) [inline doc/source/release/1.20.0-notes.rst:87]: The reference is giving an error.

File: doc/source/release/1.20.0-notes.rst [hunk 3]
```diff
@@ -75,32 +75,59 @@ Using the aliases of builtin types like ``np.int`` is deprecated
...
+``numpy.unicode``  ``str``       `numpy.unicode_`
+=================  ============  ==================================================================
+
+To give a clear guideline for the vast majority of cases, for the types
+``bool``, ``object``, ``str`` (and ``unicode``) using the plain version
+is shorter and clear, and generally a good replacement.
+For ``float`` and ``complex`` you can use ``float64`` and ``complex128``
+if you wish to be more explicit about the precision.
+
+For ``np.int`` a direct replacement with ``np.int_`` or ``int `` is also
```
Comment (charris) [inline doc/source/release/1.20.0-notes.rst:108]: Something is causing a problem here. Maybe the space after `int`?

**Comment by seberg** [comment]
Thanks, have to try how to fix that reference. Will fix (maybe tomorrow morning).  There was some discussion that this release note/deprecation is maybe a bit unclear. I hope this is a bit clearer (and "recipe style"), so the idea was to change it before the release.

File: doc/source/release/1.20.0-notes.rst [hunk 4]
```diff
@@ -75,32 +75,59 @@ Using the aliases of builtin types like ``np.int`` is deprecated
...
+=================  ============  ==================================================================
+
+To give a clear guideline for the vast majority of cases, for the types
+``bool``, ``object``, ``str`` (and ``unicode``) using the plain version
+is shorter and clear, and generally a good replacement.
+For ``float`` and ``complex`` you can use ``float64`` and ``complex128``
+if you wish to be more explicit about the precision.
+
+For ``np.int`` a direct replacement with ``np.int_`` or ``int`` is also
+good, but the precision depends on the computer and operating system.
```
Comment (eric-wieser) [inline doc/source/release/1.20.0-notes.rst:109]: Perhaps this isn't the place to say it, but I think it's important to note that `np.int_` and `int` are interpreted in the same way, even though that way is platform-dependent. This wording suggests that these replacements are platform dependent while the original was not, which is misleading.
```suggestion
good and will not change behavior, but the precision will continue to depend
on the computer and operating system.
```

File: doc/source/release/1.20.0-notes.rst [hunk 5]
```diff
@@ -75,32 +75,59 @@ Using the aliases of builtin types like ``np.int`` is deprecated
...
+When used with ``np.dtype(...)`` or ``dtype=...`` changing it to the
+NumPy name as mentioned above will have no effect on the output.
+If used as a scalar with::
+
+    np.float(123)
+
+changing it can subtly change the result.  In this case, the Python version
+``float(123)`` or ``int(12.)`` is normally preferable, although the NumPy
+version may be useful for consistency with NumPy arrays. For example, since
+NumPy behaves differently for things like division by zero.
```
Comment (eric-wieser) [inline doc/source/release/1.20.0-notes.rst:131]: Last sentence feels incomplete to me, perhaps better as:
```suggestion
version may be useful for consistency with NumPy arrays (for example,
NumPy behaves differently for things like division by zero).
```

File: numpy/__init__.py [hunk 6]
```diff
@@ -165,33 +165,58 @@
...
+        "Deprecated in NumPy 1.20; for more details and guidance: "
+        "https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations")
+
+    _specific_msg = (
+        "If you specifically wanted the numpy scalar type, use `np.{}` here.")
+
+    _int_extended_msg = (
+        "When replacing `np.{}`, you may wish to use e.g. `np.int64` "
+        "or `np.int32` to specify the precision. If you wish to review "
+        "your current use check the release note link for "
```
Comment (eric-wieser) [inline numpy/__init__.py:182]:
```suggestion
        "your current use, check the release note link for "
```

File: numpy/__init__.py [hunk 7]
```diff
@@ -165,33 +165,58 @@
...
+
+    del n, extended_msg
+
+    _msg = (
+        "`np.{n}` is a deprecated alias for `np.compat.{n}`. "
+        "To silence this warning, use `np.compat.{n}` by itself. "
+        "In the likely event your code does not need to work on Python 2 "
+        "you can use the builtin `{n2}` for which `np.compat.{n}` is itself "
+        "an alias. Doing this will not modify any behaviour and is safe. "
+        "{extended_msg}\n"
```
Comment (eric-wieser) [inline numpy/__init__.py:205]: You just `del`'d `extended_msg`, so this will fail

File: numpy/__init__.py [hunk 8]
```diff
@@ -165,33 +165,58 @@
...
+        ("float", _specific_msg.format("float64")),
+        ("complex", _specific_msg.format("complex128")),
+        ("str", _specific_msg.format("str_")),
+        ("int", _int_extended_msg.format("int"))]
+
+    for n, extended_msg in _type_info:
+        __deprecated_attrs__[n] = (getattr(_builtins, n),
+                                   _msg.format(n=n, extended_msg=extended_msg))
+
+    del n, extended_msg
```
Comment (eric-wieser) [inline numpy/__init__.py:197]: Possibly better as
```suggestion
    __deprecated_attrs__.update({
       n: (getattr(_builtins, n), _msg.format(n=n, extended_msg=extended_msg))
       for n, extended_msg in _type_info})
```
just for the sake of avoiding the `del`

**Comment by eric-wieser** [comment]
This looks great, thanks @seberg!

File: doc/source/release/1.20.0-notes.rst [hunk 9]
```diff
@@ -75,32 +75,60 @@ Using the aliases of builtin types like ``np.int`` is deprecated
...
+good and will not change behavior, but the precision will continue to depend
+on the computer and operating system.
+If you want to be more explicit and review the current use, you have the
+following alternatives:
+
+* ``np.int64`` or ``np.int32`` to specify the precision exactly.
+  This ensures that results cannot depend on the computer or operating system.
+* ``np.int_`` or ``int`` (the default), but be aware that it depends on
+  the computer and operating system.
+* The C types: ``np.cint`` (int), ``np.int_`` (long), ``np.longlong``,
```
Comment (seberg) [inline doc/source/release/1.20.0-notes.rst:118]:
```suggestion
* The C types: ``np.cint`` (int), ``np.int_`` (long), ``np.longlong``.
```

**Comment by seberg** [comment]
Thanks Eric and Ross for figuring out how to correctly link the user guide... applied the changes, so should be good to go (assuming CI passes now).

**Comment by charris** [comment]
Thanks sebastian.

**Pull request merged**